plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,317 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module PlutusCore.Generators.PIR.GenTm
+  ( module PlutusCore.Generators.PIR.GenTm
+  , module Export
+  , Arbitrary (..)
+  , Gen
+  ) where
+
+import Control.Monad.Except
+import Control.Monad.Reader
+
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Maybe
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Set.Lens (setOf)
+import Data.String
+import Test.QuickCheck (Arbitrary (..), Gen)
+import Test.QuickCheck qualified as QC
+import Test.QuickCheck.GenT as Export hiding (var)
+
+import PlutusCore.Default
+import PlutusCore.Name
+import PlutusIR
+import PlutusIR.Compiler
+import PlutusIR.Subst
+
+import PlutusCore.Generators.PIR.Common
+import PlutusCore.Generators.PIR.Utils
+
+instance MonadReader r m => MonadReader r (GenT m) where
+    ask = lift ask
+    local f (GenT k) = GenT $ \qc size -> local f $ k qc size
+
+-- | Term generators carry around a context to know
+-- e.g. what types and terms are in scope.
+type GenTm = GenT (Reader GenEnv)
+
+data GenEnv = GenEnv
+  { geSize               :: Int
+  -- ^ Generator size bound. See Note [Recursion Control and geSize]
+  , geDatas              :: Map TyName (Datatype TyName Name DefaultUni ())
+  -- ^ Datatype context
+  , geTypes              :: TypeCtx
+  -- ^ Type context
+  , geTerms              :: Map Name (Type TyName DefaultUni ())
+  -- ^ Term context
+  , geUnboundUsedTyNames :: Set TyName
+  -- ^ Names that we have generated and don't want to shadow but haven't bound yet.
+  , geEscaping           :: AllowEscape
+  -- ^ Are we in a place where we are allowed to generate a datatype binding?
+  , geCustomGen          :: Maybe (Type TyName DefaultUni ())
+                         -- TODO: this could return a `Maybe` perhaps. Or it
+                         -- could be `Maybe (Maybe Type -> ...)`.
+                         -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ())
+  -- ^ A custom user-controlled generator for terms - useful for situations where
+  -- we want to e.g. generate custom strings for coverage or test some specific
+  -- pattern that generates a special case for the compiler.
+  , geCustomFreq         :: Int
+  -- ^ How often do we use the custom generator -
+  -- values in the range of 10-30 are usually reasonable.
+  , geDebug              :: Bool
+  -- ^ Are we currently running in debug-mode (to debug our generators)
+  }
+
+{- Note [Recursion Control and geSize]
```

Comment:
:+1:

Reply:
Yeah, that one by Maximilian was super helpful.